File: java/src/gherkin/Parser.java
package gherkin;

import gherkin.lexer.En;

import java.io.*;
import java.util.*;

public class Parser implements Listener {
    private List<List<String>> transitionTable;
    private Map<String, Integer> states = new HashMap<String, Integer>();
    private Map<String, Integer> events = new HashMap<String, Integer>();
    private String state;
    private Listener listener;
    private boolean throwOnError;
    private final String machinePath;

    public Parser(Listener listener) {
        this(listener, true);
    }

    public Parser(Listener listener, boolean throwOnError) {
        this(listener, throwOnError, "root");
    }

    public Parser(Listener listener, boolean throwOnError, String state) {
        this.listener = listener;
        this.throwOnError = throwOnError;
        this.state = state;
        this.machinePath = "/gherkin/parser/" + state + ".txt";
        initializeMachine();
    }

    private void initializeMachine() {
        Listener stateMachineReader = new Listener() {
            public void tag(String name, int line) {
            }

            public void comment(String content, int line) {
            }

            public void feature(String keyword, String name, int line) {
            }

            public void background(String keyword, String name, int line) {
            }

            public void scenario(String keyword, String name, int line) {
            }

            public void scenario_outline(String keyword, String name, int line) {
            }

            public void examples(String keyword, String name, int line) {
            }

            public void step(String keyword, String name, int line) {
            }

            public void py_string(int startCol, String string, int line) {
            }

            public void syntax_error(String name, String event, List<String> strings, int line) {
            }

            public void table(List<List<String>> rows, int line) {
                transitionTable = rows;
            }
        };
        Lexer l = new En(stateMachineReader);
        try {
            l.scan(readResourceAsString(machinePath));
            initStates();
            initEvents();
        } catch (Exception e) {
            throw new RuntimeException("Couldn't read " + machinePath, e);
        }
    }

    public void tag(String name, int line) {
        if (event("tag", line))
            listener.tag(name, line);
    }

    public void py_string(int startCol, String string, int line) {
        if (event("py_string", line))
            listener.py_string(startCol, string, line);
    }

    public void feature(String keyword, String name, int line) {
        if (event("feature", line))
            listener.feature(keyword, name, line);
    }

    public void background(String keyword, String name, int line) {
        if (event("background", line))
            listener.background(keyword, name, line);
    }

    public void scenario(String keyword, String name, int line) {
        if (event("scenario", line))
            listener.scenario(keyword, name, line);
    }

    public void scenario_outline(String keyword, String name, int line) {
        if (event("scenario_outline", line))
            listener.scenario_outline(keyword, name, line);
    }

    public void examples(String keyword, String name, int line) {
        if (event("examples", line))
            listener.examples(keyword, name, line);
    }

    public void step(String keyword, String name, int line) {
        if (event("step", line))
            listener.step(keyword, name, line);
    }

    public void comment(String content, int line) {
        if (event("comment", line))
            listener.comment(content, line);
    }

    public void table(List<List<String>> rows, int line) {
        if (event("table", line))
            listener.table(rows, line);
    }

    public void syntax_error(String name, String event, List<String> strings, int line) {
    }

    private void initStates() {
        int i = 0;
        for (List<String> row : transitionTable) {
            if (i > 0) {
                states.put(row.get(0), i);
            }
            i++;
        }
    }

    private void initEvents() {
        int i = 0;
        for (String event : transitionTable.get(0)) {
            if (i > 0) {
                events.put(event, i);
            }
            i++;
        }
    }

    // Return false if there was a parse error.
    private boolean event(String event, int line) {
        int row = stateIdx(state);
        int col = eventIdx(event);
        String newState = transitionTable.get(row).get(col);
        if (newState.equals("E")) {
            if (throwOnError) {
                throw new ParseError("Parse error on line " + line + ". Found " + event + " when expecting one of: " + join(expected(), ", ") + ".");
            } else {
                listener.syntax_error(state, event, expected(), line);
                return false;
            }
        } else if (newState.equals("-")) {
        } else {
            state = newState;
        }
        return true;
    }

    private List<String> expected() {
        int row = stateIdx(state);
        List<String> expected = new ArrayList<String>();
        int i = -1;
        for (String state : transitionTable.get(row)) {
            i++;
            if (i == 0) {
                continue;
            }
            if (!state.equals("E")) {
                expected.add(transitionTable.get(0).get(i));
            }
        }
        Collections.sort(expected);
        return expected;
    }

    private int stateIdx(String state) {
        return states.get(state);
    }

    private int eventIdx(String state) {
        return events.get(state);
    }

    private static String join(List<String> strings, String separator) {
        StringBuffer sb = new StringBuffer();
        int i = 0;
        for (String s : strings) {
            if (i != 0) sb.append(separator);
            sb.append(s);
            i++;
        }
        return sb.toString();
    }

    private static String readResourceAsString(String filePath) throws IOException {
        Reader machine = new InputStreamReader(Parser.class.getResourceAsStream(filePath));

        final char[] buffer = new char[0x10000];
        StringBuilder out = new StringBuilder();
        int read = 0;
        do {
            read = machine.read(buffer, 0, buffer.length);
            if (read > 0) {
                out.append(buffer, 0, read);
            }
        } while (read >= 0);
        return out.toString();
    }
}

File: dotnet/Gherkin/Parser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace Gherkin
{
    public class Parser : IListener
    {
        private IList<Machine> machines = new List<Machine>();

        private IListener listener;
        private bool throwOnError;

        public Parser(IListener listener)
            : this(listener, true)
        {
        }

        public Parser(IListener listener, bool throwOnError)
            : this(listener, throwOnError, "root")
        {
        }

        public Parser(IListener listener, bool throwOnError, string machineName)
        {
            this.listener = listener;
            this.throwOnError = throwOnError;
            pushMachine(machineName);
        }

        private void pushMachine(string machineName)
        {
            machines.Add(new Machine(this, machineName));
        }

        private void popMachine()
        {
            machines.RemoveAt(machines.Count - 1);
        }

        public void Tag(Token name)
        {
            if (NextEvent("tag", name.Position))
                listener.Tag(name);
        }

        public void PythonString(Token pyString)
        {
            if (NextEvent("py_string", pyString.Position))
                listener.PythonString(pyString);
        }

        public void Feature(Token keyword, Token name)
        {
            if (NextEvent("feature", keyword.Position))
                listener.Feature(keyword, name);
        }

        public void Background(Token keyword, Token name)
        {
            if (NextEvent("background", keyword.Position))
                listener.Background(keyword, name);
        }

        public void Scenario(Token keyword, Token name)
        {
            if (NextEvent("scenario", keyword.Position))
                listener.Scenario(keyword, name);
        }

        public void ScenarioOutline(Token keyword, Token name)
        {
            if (NextEvent("scenario_outline", keyword.Position))
                listener.ScenarioOutline(keyword, name);
        }

        public void Examples(Token keyword, Token name)
        {
            if (NextEvent("examples", keyword.Position))
                listener.Examples(keyword, name);
        }

        public void Step(Token keyword, Token name, StepKind stepKind)
        {
            if (NextEvent("step", keyword.Position))
                listener.Step(keyword, name, stepKind);
        }

        public void Comment(Token content)
        {
            if (NextEvent("comment", content.Position))
                listener.Comment(content);
        }

        public void Row(IList<Token> row, Position tablePosition)
        {
            if (NextEvent("row", tablePosition))
                listener.Row(row, tablePosition);
        }

        public void Eof()
        {
            if (NextEvent("eof", new Position(-1,-1)))
                listener.Eof();
        }

        public void SyntaxError(string state, string @event, IEnumerable<string> legalEvents, Position position)
        {
        }

        private bool NextEvent(string @event, Position position)
        {
            try
            {
                machine().NextEvent(@event, position);
                return true;
            }
            catch (ParseException e)
            {
                if (throwOnError) throw;

                listener.SyntaxError(e.State, @event, e.ExpectedEvents, position);
                return false;
            }
        }

        private Machine machine()
        {
            return machines[machines.Count - 1];
        }

        private sealed class Machine
        {
            private static readonly Regex PUSH = new Regex("push\\((.+)\\)");
            private static readonly IDictionary<string, IDictionary<string, IDictionary<string, string>>> TRANSITION_MAPS = new Dictionary<string, IDictionary<string, IDictionary<string, string>>>();

            private readonly Parser parser;
            private readonly string name;
            private string state;
            private IDictionary<string, IDictionary<string, string>> transitionMap;

            public Machine(Parser parser, string name)
            {
                this.parser = parser;
                this.name = name;
                this.state = name;
                this.transitionMap = TransitionMap(name);
            }

            public void NextEvent(string @event, Position position)
            {
                if (!transitionMap.ContainsKey(state))
                {
                    throw new Exception("Unknown state: " + state + " for machine " + name);
                }
                var states = transitionMap[state];

                if (!states.ContainsKey(@event))
                {
                    throw new Exception("Unknown transition: " + @event + " among " + states + " for machine " + name + " in state " + state);
                }

                var newState = states[@event];

                if ("E" == newState)
                {
                    throw new ParseException(state, @event, ExpectedEvents, position);
                }

                Match push = PUSH.Match(newState);
                if (push.Success)
                {
                    parser.pushMachine(push.Groups[1].Captures[0].Value);
                    parser.NextEvent(@event, position);
                }
                else if ("pop()" == newState)
                {
                    parser.popMachine();
                    parser.NextEvent(@event, position);
                }
                else
                {
                    state = newState;
                }
            }

            private IEnumerable<string> ExpectedEvents
            {
                get
                {
                    var result = new List<string>();
                    foreach (var @event in transitionMap[state].Keys)
                    {
                        if (transitionMap[state][@event] != "E")
                        {
                            result.Add(@event);
                        }
                    }
                    result.Sort();
                    return result;
                }
            }

            private IDictionary<string, IDictionary<string, string>> TransitionMap(string name)
            {
                if (!TRANSITION_MAPS.ContainsKey(name))
                {
                    TRANSITION_MAPS[name] = BuildTransitionMap(name);
                }
                return TRANSITION_MAPS[name];
            }

            private IDictionary<string, IDictionary<string, string>> BuildTransitionMap(string name)
            {
                var result = new Dictionary<string, IDictionary<string, string>>();
                var transitionTable = new StateMachine.StateMachineReader(name).TransitionTable;
                var events = transitionTable[0].Skip(1);
                foreach (var actions in transitionTable.Skip(1))
                {
                    var transitions = new Dictionary<string, string>();
                    int col = 1;
                    foreach (var @event in events)
                    {
                        transitions[@event] = actions[col++];
                    }
                    var state = actions[0];
                    result[state] = transitions;
                }
                return result;
            }
        }
    }
}